plugins/woocommerce/src/StoreApi/Routes/V1/Agentic/CheckoutSessionsComplete.php

```diff
@@ -0,0 +1,430 @@
+		} catch ( \Exception $e ) {
+			return new \WP_REST_Response(
+				[
+					'type'         => ErrorType::PROCESSING_ERROR,
+					'code'         => ErrorCode::INVALID,
+					'content_type' => MessageContentType::PLAIN,
+					'content'      => $e->getMessage(),
+				],
+				400
+			);
+		}
+
+		/**
+		 * If payment failed, return error.
+		 */
+		if ( 'failure' === $payment_result->status || 'error' === $payment_result->status ) {
+			return new \WP_REST_Response(
+				[
+					'type'         => ErrorType::PROCESSING_ERROR,
+					'code'         => ErrorCode::PAYMENT_DECLINED,
+					'content_type' => MessageContentType::PLAIN,
+					'content'      => $payment_result->message ?? __( 'Payment was declined.', 'woocommerce' ),
+				],
+				400
+			);
+		}
+
+		/**
+		 * Store the completed order ID into the session. This will prevent new orders in this session.
+		 */
+		WC()->session->set( SessionKey::COMPLETED_ORDER_ID, $this->order->get_id() );
+
+		/**
+         * Build response from canonical cart schema with order.
+         */
+		$response_data = $this->schema->get_item_response( WC()->cart );
+		$response = rest_ensure_response( $response_data );
+
+		return AgenticCheckoutUtils::add_protocol_headers( $response, $request );
+	}
+
+	/**
+	 * Gets and formats payment request data for CheckoutTrait.
+	 *
+	 * Transforms agentic payment_data format to Store API format.
+	 *
+	 * @param \WP_REST_Request $request Request object.
+	 * @return array
+	 */
+	private function get_request_payment_data( \WP_REST_Request $request ) {
+		$payment_data = [];
+		$agentic_data = $request->get_param( 'payment_data' );
+
+		if ( ! $agentic_data ) {
+			return $payment_data;
+		}
+
+		// Transform agentic format to Store API payment_data format.
+		if ( isset( $agentic_data['token'] ) ) {
+			$payment_data['wc-agentic_commerce-token'] = wc_clean( $agentic_data['token'] );
+		}
+
+		if ( isset( $agentic_data['provider'] ) ) {
+			$payment_data['wc-agentic_commerce-provider'] = wc_clean( $agentic_data['provider'] );
+		}
+
+		return $payment_data;
+	}
+
+	/**
+	 * Gets the chosen payment method (gateway) ID for CheckoutTrait.
+	 *
+	 * @param \WP_REST_Request $request Request object.
+	 * @return string
+	 * @throws RouteException If no payment gateway is available.
+	 */
+	private function get_request_payment_method_id( \WP_REST_Request $request ) {
+		$available_gateways = WC()->payment_gateways()->get_available_payment_gateways();
+
+		if ( empty( $available_gateways ) ) {
+			throw new RouteException(
+				'woocommerce_checkout_session_no_payment_gateway_available',
+				esc_html__( 'No payment gateway available.', 'woocommerce' ),
+				400
+			);
+		}
+
+		// Look for gateway with agentic_commerce capability.
+		foreach ( $available_gateways as $gateway ) {
+			if ( $gateway->supports( 'agentic_commerce' )
+				&& method_exists( $gateway, 'get_agentic_commerce_provider' )
+				&& method_exists( $gateway, 'get_agentic_commerce_payment_methods' )
```

Comment:
Should we also verify that the gateway supports the agentic commerce provider specified in the request? It feels like we might need the return value to specify a list of supported agentic commerce providers, not just one.

Reply:
I would also want that we can give a list of providers than that just a single one. But right now, this is coupled with [the specs](https://github.com/agentic-commerce-protocol/agentic-commerce-protocol/blob/6faad74ed29e443b198b35ef05adfd806b9063fd/spec/json-schema/schema.agentic_checkout.json#L79-L82) as it allows a single provider.

Reply:
@htdat and I discussed this in a call, and this is probably over-complicating things. It almost certainly makes more sense for each gateway instance to support only one agentic commerce provider and one set of payment methods for that provider. If a particular plugin supports multiple agentic commerce providers, they can create dedicated payment gateways for each supported provider.